### ISSUE TITLE
Projection Merge will re-evaluate expensive expressions

### ISSUE THREAD

**Issue by wenleix** [prestodb/presto#7787]
I think there is an opening issue but didn't find it. So open a new one. We can mark this as duplicate and link it to that one. 

```
presto:> explain(type distributed)
        -> SELECT x+1, x+2
        -> FROM (
        ->     SELECT CAST(JSON_EXTRACT_SCALAR(x, '$.val') AS BIGINT) AS x
        ->     FROM (VALUES '{"val": 4}') AS tmp(x)
        -> )
        ->
        -> ;
                                                     Query Plan
--------------------------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]
     Output layout: [expr_5, expr_6]
     Output partitioning: SINGLE []
     - Output[_col0, _col1] => [expr_5:bigint, expr_6:bigint]
             _col0 := expr_5
             _col1 := expr_6
         - Project[] => [expr_5:bigint, expr_6:bigint]
                 expr_5 := (CAST("json_extract_scalar"("field", CAST('$.val' AS jsonpath)) AS bigint) + BIGINT '1')
                 expr_6 := (CAST("json_extract_scalar"("field", CAST('$.val' AS jsonpath)) AS bigint) + BIGINT '2')
             - LocalExchange[ROUND_ROBIN] () => field:varchar(10)
                 - Values => [field:varchar(10)]
                         ('{"val": 4}')
```

Current workaround:

```
presto:> EXPLAIN (type distributed)
        ->
        -> SELECT x+1, x+2, x+3
        -> FROM (
        ->     SELECT
        ->         IF (rand()> -1, CAST(JSON_EXTRACT_SCALAR(x, '$.val') AS BIGINT)) AS x
        ->     FROM (VALUES '{"val": 4}') AS tmp(x)
        -> )
        -> ;
                                                                   Query Plan
----------------------------------------------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]
     Output layout: [expr_5, expr_6, expr_7]
     Output partitioning: SINGLE []
     - Output[_col0, _col1, _col2] => [expr_5:bigint, expr_6:bigint, expr_7:bigint]
             _col0 := expr_5
             _col1 := expr_6
             _col2 := expr_7
         - Project[] => [expr_5:bigint, expr_6:bigint, expr_7:bigint]
                 expr_5 := ("expr_1" + BIGINT '1')
                 expr_6 := ("expr_1" + BIGINT '2')
                 expr_7 := ("expr_1" + BIGINT '3')
             - Project[] => [expr_1:bigint]
                     expr_1 := (CASE WHEN ("rand"() > -1.0) THEN CAST("json_extract_scalar"("field", CAST('$.val' AS jsonpath)) AS bigin
                 - LocalExchange[ROUND_ROBIN] () => field:varchar(10)
                     - Values => [field:varchar(10)]
                             ('{"val": 4}')
```

**Comment by martint** [comment]
This probably fixes it: https://github.com/prestodb/presto/pull/7094

**Comment by wenleix** [comment]
This is fixed in 0.174:

```
presto> explain(type distributed)
            ->          SELECT x+1, x+2
            ->          FROM (
            ->              SELECT CAST(JSON_EXTRACT_SCALAR(x, '$.val') AS BIGINT) AS x
            ->              FROM (VALUES '{"val": 4}') AS tmp(x)
            ->          )
            -> ;
                                               Query Plan
---------------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]
     Output layout: [expr_5, expr_6]
     Output partitioning: SINGLE []
     - Output[_col0, _col1] => [expr_5:bigint, expr_6:bigint]
             _col0 := expr_5
             _col1 := expr_6
         - Project[] => [expr_5:bigint, expr_6:bigint]
                 expr_5 := ("expr_1" + BIGINT '1')
                 expr_6 := ("expr_1" + BIGINT '2')
             - Project[] => [expr_1:bigint]
                     expr_1 := CAST("json_extract_scalar"("field", CAST('$.val' AS jsonpath)) AS bigint)
                 - LocalExchange[ROUND_ROBIN] () => field:varchar(10)
                     - Values => [field:varchar(10)]
                             ('{"val": 4}')

(1 row)
```
